.changeset/twelve-grapes-dream.md

```diff
@@ -0,0 +1,5 @@
+---
+'@astrojs/cloudflare': minor
+---
+
+Fix 404 responses in "exclude" routing strategy
```

Comment:
```suggestion
Fixes an issue where `404.astro` was not used.
```
Routing strategy is not directly involved, is it?
It only hid the issue when the platform could serve the prerendered `404.html` instead.

Reply:
The other issues still persist in the include strategy. This appears to be due to the routes.json issues, so this may solve that once those are fixed, but it would require further testing.

Reply:
That can be done with this npm version: `@astrojs/cloudflare@experimental--404-fix`. I'll try it in a bit.

Reply:
The branch named `latest` on your repo: https://latest.xyz-col.pages.dev/
The branch named `latest-prerendered` on your repo: https://latest-prerendered.xyz-col.pages.dev/

Both are running your fix.
Not sure about what to look for after changing the routing strategy.

Reply:
After looking into it further myself, the issues with rendering 404 pages in the include strategy aren't actually related to the routes, but how Cloudflare treats apps without a 404.html. 

Before the change that implemented include mode, (withastro/astro#7846), the default behavior was similar to what exclude mode is now.  
In include mode, all non-defined routes use Cloudflare's static method. In this method, if there is no 404.html, then it's assumed that you're running a SPA and renders `index.html` instead (with status code 200). If the index route is server rendered, it instead renders an empty 404 error.
In exclude, all routes, excluding static files, are rendered through Astro. This is what allows not-found routes to be caught by Astro and rendered appropriately.

alexanderniebuhr actually found this out yesterday when starting to figure out #64 and posted a discord thread that lead to that conclusion.

Reply:
So to conclude, this PR would not be able to change how 404 responses are resolved in include mode as missing routes are not sent through Astro.